src/_pytest/unittest.py

```diff
@@ -187,29 +191,45 @@ def addSuccess(self, testcase):
     def stopTest(self, testcase):
         pass
 
-    def _handle_skip(self):
-        # implements the skipping machinery (see #2137)
-        # analog to pythons Lib/unittest/case.py:run
+    def runtest(self):
         testMethod = getattr(self._testcase, self._testcase._testMethodName)
-        if getattr(self._testcase.__class__, "__unittest_skip__", False) or getattr(
-            testMethod, "__unittest_skip__", False
-        ):
-            # If the class or method was skipped.
-            skip_why = getattr(
-                self._testcase.__class__, "__unittest_skip_why__", ""
-            ) or getattr(testMethod, "__unittest_skip_why__", "")
-            self._testcase._addSkip(self, self._testcase, skip_why)
-            return True
-        return False
 
-    def runtest(self):
-        if self.config.pluginmanager.get_plugin("pdbinvoke") is None:
+        class _GetOutOf_testPartExecutor(KeyboardInterrupt):
+            """Helper exception to get out of unittests's testPartExecutor."""
+
+        unittest = sys.modules.get("unittest")
+
+        reraise = ()
+        if unittest:
+            reraise += (unittest.SkipTest,)
+
+        @functools.wraps(testMethod)
+        def wrapped_testMethod(*args, **kwargs):
+            try:
+                self.ihook.pytest_pyfunc_call(pyfuncitem=self)
+            except reraise:
+                raise
+            except Exception as exc:
+                expecting_failure_method = getattr(
+                    testMethod, "__unittest_expecting_failure__", False
+                )
+                expecting_failure_class = getattr(
+                    self, "__unittest_expecting_failure__", False
+                )
+                expecting_failure = expecting_failure_class or expecting_failure_method
+                self._need_tearDown = True
+
+                if expecting_failure:
+                    raise
+
+                raise _GetOutOf_testPartExecutor(exc)
+
+        self._testcase._wrapped_testMethod = wrapped_testMethod
```

Comment:
I think we should override the original method, otherwise the internal name will show up to the user when entering PDB:

```
self = <foo.T testMethod=_wrapped_testMethod>

    def test_foo(self):
        self.filename = 'hello'
>       assert 0
```